File: python/src/ancilis/middleware/discovery.py
"""Auto-discovery: registers tools from MCP list_tools into the ToolRegistry."""

from __future__ import annotations

import hashlib
import logging
from dataclasses import dataclass
from typing import Any

from ancilis.engine.registry import ToolEntry, ToolRegistry, ToolStatus

logger = logging.getLogger("ancilis.discovery")


@dataclass
class DriftEvent:
    tool_name: str
    old_hash: str | None
    new_hash: str


def register_tools_from_list(
    tools: list[Any],
    registry: ToolRegistry,
    pre_approved: list[str] | None = None,
) -> list[DriftEvent]:
    """Register tools from MCP list_tools response. Returns drift events.

    Tools are registered as OBSERVED by default. Tools listed in
    pre_approved (from security.tools.allowed) are registered as APPROVED.
    """
    drift_events: list[DriftEvent] = []
    approved_set = set(pre_approved) if pre_approved else set()

    for tool in tools:
        name: str = tool.name
        description: str = tool.description or ""
        desc_hash = hashlib.sha256(description.encode()).hexdigest()

        existing = registry.lookup(name)
        if existing and existing.description_hash and existing.description_hash != desc_hash:
            drift_events.append(DriftEvent(
                tool_name=name,
                old_hash=existing.description_hash,
                new_hash=desc_hash,
            ))
            logger.warning(
                "Tool description drift detected for '%s': hash changed from %s to %s",
                name, existing.description_hash[:12], desc_hash[:12],
            )

        # Pre-approved tools (from config allowlist) get APPROVED status;
        # all others start as OBSERVED until explicitly approved.
        status = ToolStatus.APPROVED if name in approved_set else ToolStatus.OBSERVED

        registry.register(ToolEntry(
            name=name,
            description_hash=desc_hash,
            status=status,
        ))

    return drift_events

File: typescript/src/ancilis/middleware/discovery.ts
/** Auto-discovery: registers tools from MCP listTools into the ToolRegistry. */

import { createHash } from "node:crypto";
import { type ToolRegistry, ToolStatus } from "../engine/registry.js";

export interface DriftEvent {
  toolName: string;
  oldHash: string | null;
  newHash: string;
}

export interface DiscoverableTool {
  name: string;
  description?: string | null;
}

export function registerToolsFromList(
  tools: DiscoverableTool[],
  registry: ToolRegistry,
  preApproved: string[] = [],
): DriftEvent[] {
  const driftEvents: DriftEvent[] = [];
  const approvedSet = new Set(preApproved);

  for (const tool of tools) {
    const description = tool.description ?? "";
    const descHash = createHash("sha256").update(description).digest("hex");

    const existing = registry.lookup(tool.name);
    if (existing?.descriptionHash && existing.descriptionHash !== descHash) {
      driftEvents.push({
        toolName: tool.name,
        oldHash: existing.descriptionHash,
        newHash: descHash,
      });
    }

    // Don't downgrade an already-approved tool
    if (existing && existing.status === ToolStatus.APPROVED) {
      // Update hash only
      registry.register({
        ...existing,
        descriptionHash: descHash,
      });
    } else {
      const now = new Date().toISOString();
      registry.register({
        name: tool.name,
        descriptionHash: descHash,
        status: approvedSet.has(tool.name) ? ToolStatus.APPROVED : ToolStatus.OBSERVED,
        approvedBy: approvedSet.has(tool.name) ? "config" : null,
        firstSeen: existing?.firstSeen ?? now,
        statusChanged: now,
      });
    }
  }

  return driftEvents;
}
